File: Algorithms/src/3464. Maximize the Distance Between Points on a Square/Code/solution.js
/**
 * @param {number} side
 * @param {number[][]} points
 * @param {number} k
 * @return {number}
 */
var maxDistance = function (side, points, k) {
  const pts = points.map(([x, y]) => {
    let pos;
    if (x === 0) pos = y;
    else if (y === side) pos = side + x;
    else if (x === side) pos = 3 * side - y;
    else pos = 4 * side - x;
    return { pos, x, y };
  });

  pts.sort((a, b) => a.pos - b.pos);

  function getOffset(x, y, d) {
    if (x === 0) {
      if (d <= 2 * side - y) return d;
      if (d <= side + y) return 2 * side + d - 2 * y;
      return -1;
    } else if (y === side) {
      if (d <= 2 * side - x) return d;
      if (d <= side + x) return 2 * side + d - 2 * x;
      return -1;
    } else if (x === side) {
      if (d <= side + y) return d;
      if (d <= 2 * side - y) return d + 2 * y;
      return -1;
    } else {
      if (d <= side + x) return d;
      if (d <= 2 * side - x) return d + 2 * x;
      return -1;
    }
  }

  function lowerBound(arr, l, r, target) {
    while (l < r) {
      const m = (l + r) >> 1;
      if (arr[m] < target) l = m + 1;
      else r = m;
    }
    return l;
  }

  function can(d) {
    const n = pts.length;
    const pos3 = new Array(3 * n);
    for (let i = 0; i < n; i++) {
      pos3[i] = pts[i].pos;
      pos3[i + n] = pts[i].pos + 4 * side;
      pos3[i + 2 * n] = pts[i].pos + 8 * side;
    }

    const nxt = new Array(2 * n).fill(-1);

    for (let i = 0; i < 2 * n; i++) {
      const p = pts[i % n];
      const off = getOffset(p.x, p.y, d);
      if (off < 0) continue;

      const target = pos3[i] + off;
      const hi = Math.min(i + n, 3 * n);
      const j = lowerBound(pos3, i + 1, hi, target);
      if (j < hi) nxt[i] = j;
    }

    for (let start = 0; start < n; start++) {
      let cur = start;
      let cnt = 1;

      while (cnt < k) {
        cur = nxt[cur];
        if (cur === -1 || cur >= start + n) break;
        cnt++;
      }

      if (cnt >= k) {
        const a = pts[start];
        const b = pts[cur % n];
        if (Math.abs(a.x - b.x) + Math.abs(a.y - b.y) >= d) {
          return true;
        }
      }
    }

    return false;
  }

  let lo = 0,
    hi = 2 * side;
  while (lo < hi) {
    const mid = Math.floor((lo + hi + 1) / 2);
    if (can(mid)) lo = mid;
    else hi = mid - 1;
  }

  return lo;
};

File: Algorithms/src/3464. Maximize the Distance Between Points on a Square/Code/solution.java
import java.util.*;

class Solution {
    static class P {
        long pos, x, y;

        P(long pos, long x, long y) {
            this.pos = pos;
            this.x = x;
            this.y = y;
        }
    }

    private long getOffset(long side, long x, long y, long d) {
        if (x == 0) { // left
            if (d <= 2L * side - y)
                return d;
            if (d <= side + y)
                return 2L * side + d - 2L * y;
            return -1;
        } else if (y == side) { // top
            if (d <= 2L * side - x)
                return d;
            if (d <= side + x)
                return 2L * side + d - 2L * x;
            return -1;
        } else if (x == side) { // right
            if (d <= side + y)
                return d;
            if (d <= 2L * side - y)
                return d + 2L * y;
            return -1;
        } else { // bottom
            if (d <= side + x)
                return d;
            if (d <= 2L * side - x)
                return d + 2L * x;
            return -1;
        }
    }

    private int lowerBound(long[] arr, int l, int r, long target) {
        while (l < r) {
            int m = (l + r) >>> 1;
            if (arr[m] < target)
                l = m + 1;
            else
                r = m;
        }
        return l;
    }

    private boolean can(long side, P[] pts, int k, long d) {
        int n = pts.length;

        long[] pos3 = new long[3 * n];
        for (int i = 0; i < n; i++) {
            pos3[i] = pts[i].pos;
            pos3[i + n] = pts[i].pos + 4L * side;
            pos3[i + 2 * n] = pts[i].pos + 8L * side;
        }

        int[] nxt = new int[2 * n];
        Arrays.fill(nxt, -1);

        for (int i = 0; i < 2 * n; i++) {
            P p = pts[i % n];
            long off = getOffset(side, p.x, p.y, d);
            if (off < 0)
                continue;

            long target = pos3[i] + off;
            int hi = Math.min(i + n, 3 * n);
            int j = lowerBound(pos3, i + 1, hi, target);
            if (j < hi)
                nxt[i] = j;
        }

        for (int start = 0; start < n; start++) {
            int cur = start;
            int cnt = 1;

            while (cnt < k) {
                cur = nxt[cur];
                if (cur == -1 || cur >= start + n)
                    break;
                cnt++;
            }

            if (cnt >= k) {
                long dx = Math.abs(pts[start].x - pts[cur % n].x);
                long dy = Math.abs(pts[start].y - pts[cur % n].y);
                if (dx + dy >= d)
                    return true;
            }
        }

        return false;
    }

    public int maxDistance(int side, int[][] points, int k) {
        P[] pts = new P[points.length];

        for (int i = 0; i < points.length; i++) {
            long x = points[i][0];
            long y = points[i][1];
            long pos;
            if (x == 0)
                pos = y;
            else if (y == side)
                pos = 1L * side + x;
            else if (x == side)
                pos = 3L * side - y;
            else
                pos = 4L * side - x;
            pts[i] = new P(pos, x, y);
        }

        Arrays.sort(pts, Comparator.comparingLong(a -> a.pos));

        long lo = 0, hi = 2L * side;
        while (lo < hi) {
            long mid = (lo + hi + 1) >>> 1;
            if (can(side, pts, k, mid))
                lo = mid;
            else
                hi = mid - 1;
        }

        return (int) lo;
    }
}